odoo/addons/base/models/ir_mail_server.py

```diff
@@ -491,20 +525,31 @@ def send_email(self, message, mail_server_id=None, smtp_server=None, smtp_port=N
         :return: the Message-ID of the message that was just sent, if successfully sent, otherwise raises
                  MailDeliveryException and logs root cause.
         """
-        smtp_from, smtp_to_list, message = self._prepare_email_message(message)
+        smtp = smtp_session
+        if not smtp:
+            smtp = self.connect(
+                smtp_server, smtp_port, smtp_user, smtp_password,
+                smtp_encryption, smtp_debug, mail_server_id=mail_server_id, smtp_from=message['From'],
+                ssl_certificate=smtp_ssl_certificate, ssl_private_key=smtp_ssl_private_key)
+
+        smtp_from, smtp_to_list, message = self._prepare_email_message(message, smtp)
 
         # Do not actually send emails in testing mode!
         if getattr(threading.currentThread(), 'testing', False) or self.env.registry.in_test_mode():
             _test_logger.info("skip sending email in test mode")
+            if isinstance(smtp, tools.TestingSMTPSession):
```

Comment:
Still some test-related code to move into proper mocks and tools (makes code quite obfuscated to understand).

Reply:
Ok, I'll try to remove it and to make custom mock methods :)